[pytorch] process_group_agent optimizations

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29324 [pytorch] process_group_agent optimizations**

This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

Commit:
[pytorch] process_group_agent optimizations

This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Question: if we switch from `torch::save`/`torch::load` to jit `pickle`/`unpickle`, do we still need to explicitly mark the serialization mode?

Commit:
Update on "[pytorch] process_group_agent optimizations"


This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

[ghstack-poisoned]

jjlilley commented:
> Question: if we switch from torch::save/torch::load to jit pickle/unpickle, do we still need to explicitly
> mark the serialization mode?

If we decide that there's a better pickling mechanism, we can always get rid of this.

My main goal here was that in case different sorts of encoding payloads have more optimal encoding schemes, that we have an out-of-band (and effectively free) mechanism to specify this.

Right now, if we have a minimally-sized response from an RPC (e.g. an ACK), without any tensors, we expend more effort/bytes encoding this than we need to.

Being able to alternate between torch::save / jit::pickle / payload-only / whatever, depending on the input, seems like it could help some cases.

Commit:
Update on "[pytorch] process_group_agent optimizations"


This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] process_group_agent optimizations"


This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] process_group_agent optimizations"


This change adds a few optimizations to process_group_agent:

 1) Don't add an extra tensor during serialization for the message id,
    but instead put that in the preamble tensor. This saves maybe 15%
    overhead for minimal-sized RPCs.

 2) Add a payload-only fastpath.
    a) For very tiny messages (e.g. acks, status updates), this reduces RPC overhead by roughly 50%, by removing zip file overhead and related setup/copying in torch::load()/torch::save().
    b) If we do end up with large non-tensor payloads, this saves us ~25% in benchmarks, from avoiding copying in torch::save/load.

Differential Revision: [D18352261](https://our.internmc.facebook.com/intern/diff/D18352261/)

[ghstack-poisoned]

jjlilley commented:
Closing this PR for a bit, there's another approach that I'd like to take...
